libraries/lib-module-manager/PluginHost.cpp

```diff
@@ -0,0 +1,220 @@
+#include "PluginManager.h"
+
+namespace
+{
+   //Attempts to instantiate plugin module and put plugin descriptors into result
+   void Discover(details::PluginValidationResult& result, const wxString& providerId, const wxString& pluginPath)
+   {
+      try
+      {
+         if(auto provider = ModuleManager::Get().CreateProviderInstance(providerId, wxEmptyString))
+         {
+            TranslatableString errorMessage{};
+
+            provider->DiscoverPluginsAtPath(pluginPath, errorMessage, [&result](PluginProvider *provider, ComponentInterface *ident)
+            {
+               //Workaround: use DefaultRegistrationCallback to create all descriptors for us
+               //and then put a copy into result
+               auto id = PluginManager::DefaultRegistrationCallback(provider, ident);
+               if(const auto desc = PluginManager::Get().GetPlugin(id))
+                  result.Add(PluginDescriptor { *desc });
+               return id;
+            });
+            if(!errorMessage.empty())
+               result.SetError(errorMessage.Translation());
+         }
+         else
+            result.SetError("provider not found");
+      }
+      catch(...)
+      {
+         result.SetError("unknown error");
+      }
+   }
+}
+
+PluginHost::PluginHost()
+{
+   FileNames::InitializePathList();
+
+   wxFileName configFileName{ FileNames::Configuration() };
+   auto pConfig = std::make_unique<FileConfig>(
+      AppName, wxEmptyString, configFileName.GetFullPath(),
+      wxEmptyString, wxCONFIG_USE_LOCAL_FILE);
+   pConfig->Init();
+   InitPreferences(std::move(pConfig));
+
+   auto& moduleManager = ModuleManager::Get();
+   moduleManager.Initialize();
+   moduleManager.DiscoverProviders();
+
+   mClient = std::make_unique<IPCClient>(*this);
+}
+
+void PluginHost::OnConnect(IPCChannel& channel) noexcept
+{
+   std::lock_guard lck(mSync);
+   mChannel = &channel;
+}
+
+void PluginHost::OnDisconnect() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnConnectionError() noexcept
+{
+   Stop();
+}
+
+void PluginHost::OnDataAvailable(const void* data, size_t size) noexcept
+{
+   try
+   {
+      mInputMessageReader.ConsumeBytes(data, size);
+      if(mInputMessageReader.CanPop())
+      {
+         {
+            std::lock_guard lck(mSync);
+            assert(!mRequest);
+            mRequest = mInputMessageReader.Pop();
+         }
+         mRequestCondition.notify_one();
+      }
+   }
+   catch(...)
+   {
+      Stop();
+   }
+}
+
+bool PluginHost::Serve()
+{
+   std::unique_lock lck(mSync);
+   mRequestCondition.wait(lck, [this]{ return !mRunning || mRequest.has_value(); });
+
+   if(!mRunning)
+      return false;
+
+   if(mRequest)
+   {
+      std::optional<wxString> request;
+      mRequest.swap(request);
+
+      lck.unlock();
+      
+      wxString providerId;
+      wxString pluginPath;
+      details::PluginValidationResult result;
+      if(details::ParseRequestString(*request, providerId, pluginPath))
```

Comment:
This implies that this server expects messages as formatted by `MakeRequestString`.  Then that should be commented in the header file.